File: backend/src/main/java/com/laras/service/TestimonialService.java
package com.laras.service;

import com.laras.dto.TestimonialDto;
import com.laras.entity.Testimonial;
import com.laras.repository.TestimonialRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class TestimonialService {

    private final TestimonialRepository testimonialRepository;

    public List<TestimonialDto> getAllTestimonials() {
        return testimonialRepository.findAllByOrderByDisplayOrderAsc()
                .stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    public List<TestimonialDto> getVisibleTestimonials() {
        return testimonialRepository.findByIsVisibleTrueOrderByDisplayOrderAsc()
                .stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    public TestimonialDto getTestimonialById(Long id) {
        Testimonial testimonial = testimonialRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Testimonial not found with id: " + id));
        return toDto(testimonial);
    }

    public TestimonialDto createTestimonial(TestimonialDto dto) {
        Testimonial testimonial = Testimonial.builder()
                .customerName(dto.getCustomerName())
                .text(dto.getText())
                .rating(dto.getRating() != null ? dto.getRating() : 5)
                .isVisible(dto.getIsVisible() != null ? dto.getIsVisible() : true)
                .displayOrder(dto.getDisplayOrder() != null ? dto.getDisplayOrder() : 0)
                .build();

        Testimonial saved = testimonialRepository.save(testimonial);
        return toDto(saved);
    }

    public TestimonialDto updateTestimonial(Long id, TestimonialDto dto) {
        Testimonial testimonial = testimonialRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Testimonial not found with id: " + id));

        testimonial.setCustomerName(dto.getCustomerName());
        testimonial.setText(dto.getText());
        if (dto.getRating() != null)
            testimonial.setRating(dto.getRating());
        if (dto.getIsVisible() != null)
            testimonial.setIsVisible(dto.getIsVisible());
        if (dto.getDisplayOrder() != null)
            testimonial.setDisplayOrder(dto.getDisplayOrder());

        Testimonial saved = testimonialRepository.save(testimonial);
        return toDto(saved);
    }

    public void deleteTestimonial(Long id) {
        testimonialRepository.deleteById(id);
    }

    private TestimonialDto toDto(Testimonial testimonial) {
        return TestimonialDto.builder()
                .id(testimonial.getId())
                .customerName(testimonial.getCustomerName())
                .text(testimonial.getText())
                .rating(testimonial.getRating())
                .isVisible(testimonial.getIsVisible())
                .displayOrder(testimonial.getDisplayOrder())
                .createdAt(testimonial.getCreatedAt())
                .updatedAt(testimonial.getUpdatedAt())
                .build();
    }
}

File: frontend/src/services/testimonialService.ts
import api from './api';

export interface Testimonial {
    id: number;
    customerName: string;
    text: string;
    rating: number;
    isVisible: boolean;
    displayOrder: number;
    createdAt: string;
    updatedAt: string;
}

export interface TestimonialInput {
    customerName: string;
    text: string;
    rating?: number;
    isVisible?: boolean;
    displayOrder?: number;
}

export const testimonialService = {
    getVisibleTestimonials: async (): Promise<Testimonial[]> => {
        const response = await api.get<Testimonial[]>('/testimonials');
        return response.data;
    },

    getAllTestimonials: async (): Promise<Testimonial[]> => {
        const response = await api.get<Testimonial[]>('/admin/testimonials');
        return response.data;
    },

    createTestimonial: async (data: TestimonialInput): Promise<Testimonial> => {
        const response = await api.post<Testimonial>('/admin/testimonials', data);
        return response.data;
    },

    updateTestimonial: async (id: number, data: TestimonialInput): Promise<Testimonial> => {
        const response = await api.put<Testimonial>(`/admin/testimonials/${id}`, data);
        return response.data;
    },

    deleteTestimonial: async (id: number): Promise<void> => {
        await api.delete(`/admin/testimonials/${id}`);
    },
};

export default testimonialService;
